src/main/java/com/iexec/worker/PingService.java

```diff
@@ -18,36 +18,64 @@
 
 import com.iexec.worker.config.CoreConfigurationService;
 import com.iexec.worker.feign.CustomCoreFeignClient;
+import com.iexec.worker.utils.AsyncUtils;
+import com.iexec.worker.utils.ExecutorUtils;
 import com.iexec.worker.worker.WorkerService;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Service;
 
 import java.time.LocalTime;
+import java.util.concurrent.Executor;
 
 @Slf4j
 @Service
 public class PingService {
 
+    private static final int PING_RATE = 10000; // 10s
+
+    private final Executor executor;
     private final CustomCoreFeignClient customCoreFeignClient;
     private final CoreConfigurationService coreConfigurationService;
     private final WorkerService workerService;
 
     public PingService(CustomCoreFeignClient customCoreFeignClient,
                        CoreConfigurationService coreConfigurationService,
                        WorkerService workerService) {
+        executor = ExecutorUtils
+                .newSingleThreadExecutorWithFixedSizeQueue(1, "ping-");
         this.customCoreFeignClient = customCoreFeignClient;
         this.coreConfigurationService = coreConfigurationService;
         this.workerService = workerService;
     }
 
-    @Scheduled(fixedRate = 10000)
-    public void pingScheduler() {
+    /**
+     * Trigger the scheduler ping every t seconds. The method that pings the
+     * scheduler runs asynchronously inside a new thread to liberate the thread
+     * used for @Scheduled tasks.
+     * We use single thread executor to make sure the worker does not ping more
+     * than once at the same time. The executors queue is of size 1 to avoid memory
+     * leak if the thread halts for any reason.
+     */
+    @Scheduled(fixedRate = PING_RATE)
+    void triggerSchedulerPing() {
+        log.debug("Triggering scheduler ping action");
+        AsyncUtils.runAsyncTask("ping", () -> pingScheduler(), executor);
+    }
+
+    /**
+     * Send ping message to the scheduler and save the session id when pinging for
+     * the first time. If the session id changes, it means that the scheduler has
+     * restarted, os the worker needs to restart also.
+     */
+    void pingScheduler() {
+        log.debug("Sending ping to scheduler");
         String sessionId = customCoreFeignClient.ping();
-        // log once in an hour
-        if (LocalTime.now().getMinute() == 0) {
-            log.info("Sent ping to scheduler " + sessionId);
+        // Log once in an hour, in the first ping of the first minute.
+        if (LocalTime.now().getMinute() == 0
+                && LocalTime.now().getSecond() <= PING_RATE) {
```

Comment:
Not sure about this part. It is always true as `LocalTime.now().getSecond()` will be an integer between 0 and 60 while `PING_RATE` is always 10000. You may want to check `LocalTime.now().getSecond() * 1000 <= PING_RATE` so that you use the same unit (ms) for both values.

Reply:
My bad, thanks for the heads up!